torch/fx/subgraph_rewriter.py

```diff
@@ -0,0 +1,308 @@
+                    else:
+                        # If we don't have an output-output match, we
+                        # need to keep `original_graph`'s existing
+                        # "output" node. We update the existing output
+                        # node's args and remove it from
+                        # `nodes_to_delete`
+                        assert len(n.all_input_nodes) == 1
+                        n_input = next(iter(n.all_input_nodes))
+                        new_args = [arg for arg in subgraph_output.args
+                                    if arg not in nodes_to_delete]
+                        delete_args = [arg for arg in subgraph_output.args
+                                       if arg in nodes_to_delete]
+                        new_args.append(copied_nodes[n_input])
+                        new_kwargs = subgraph_output.kwargs
+                        subgraph_output._update_args_kwargs(tuple(new_args), new_kwargs)
+                        nodes_to_delete.remove(subgraph_output)
+                        for arg in delete_args:
+                            original_graph.erase_node(arg)
+                            nodes_to_delete.remove(arg)
+                elif n.op == "placeholder":
+                    # If `n` corresponds to a placeholder node that
+                    # already exists in `original_graph`, we don't want
+                    # to swap it out
+                    if n.target in original_graph_inputs.keys():
+                        original_graph_node = original_graph_inputs[n.target]
+                        if original_graph_node in nodes_to_delete:
+                            nodes_to_delete.remove(original_graph_node)
+                        copied_nodes[n] = original_graph_node
+                    else:
+                        # Temporarily change the insertion point to the
+                        # beginning of the graph to ensure that placeholder
+                        # nodes are inserted before anything else
+                        with original_graph.inserting_after(last_placeholder):
+                            copied_nodes[n] = original_graph.node_copy(n,
+                                                                       lambda n :
+                                                                       copied_nodes[n])
+                            if n.name in original_graph_inputs.keys():
+                                original_node = original_graph_inputs[n.name]
+                                nodes_to_delete.add(original_node)
+                                original_node.replace_all_uses_with(copied_nodes[n])
+                            last_placeholder = copied_nodes[n]
+                else:
+                    copied_nodes[n] = original_graph.node_copy(n,
+                                                               lambda n :
+                                                               copied_nodes[n])
```

Comment:
The code to copy in the replacement graph from 216-278 is too complicated for what it's doing. Here's what I think we should do at a high level:

```
# Map nodes in the replacement graph to nodes in the original graph
val_map : [Node, Node] = {}

# Initialize val_map with mappings from replacement graph placeholders to original graph values
# When these values are populated, graph_copy will correctly make the copied-in nodes point
# to these nodes
for node in replacement_graph.nodes:
  if node.op == 'placeholder':
    # Going to have to figure out how to keep track of replacement graph nodes inputs -> original graph values
    # Probably going to have to:
    # 1. Keep track of pattern placeholder -> original graph nodes in _match_nodes
    # 2. Create positional correspondences between the replacement graph placeholders and the pattern placeholder
    # 3. Fetch the original graph nodes using information from (1) and (2)
    val_map[node] = ...

# Copy the replacement graph over
with original_graph.inserting_before(subgraph_output):
  output_val = original_graph.graph_copy(replacement_graph, lambda n: val_map[n])

# output_val contains the values the correspond to the new inserted replacement graph. We need to 
# go through and replace the value in the original graph with this new value
assert isinstance(output_val, Node)
output_val = output_vals[0]

pattern_graph_output_val = [node for node in pattern_graph if node.op == 'output'][0].args[0]
assert isinstance(pattern_graph_output_val, Node)

match.nodes_map[pattern_graph_output_val].replace_all_uses_with(output_val)

# Finally, eliminate dead code. This will delete the nodes from the match we just replaced
for node in reversed(original_graph.nodes):
  if len(node.users) == 0:
    original_graph.erase_node(node)
```

Note that there's a bunch of stuff that we need to keep track of here:

1) Correspondence between Placeholders in the original graph, the pattern graph, and the replacement graph. I think here we should do the following:
  a) Keep track of pattern Placeholder -> original graph values in `nodes_map` in `_match_nodes`
  b) Enforce that all pattern Placeholder values are used in the Pattern. It's pretty confusing allowing unused placeholders (e.g. in `test_subgraph_rewriter_single_pattern_match`). Essentially what happens is that any value could alias any other value in the pattern
  b) Enforce that the pattern graph has the exact same number of Placeholders as the replacement Graph. This will make it easier to keep track of correspondences when we do the above^ algorithm. Additionally, adding extra inputs to the graph is pretty surprising, so I don't think we want to keep that
2) Correspondence between the pattern output values and the original graph output values. I hacked up a few lines above to figure that out but I think we could do it more systematically